contracts/JBV3Token.sol

```diff
@@ -0,0 +1,401 @@
+
+import '@openzeppelin/contracts/token/ERC20/extensions/draft-ERC20Permit.sol';
+import '@jbx-protocol-v2/contracts/interfaces/IJBController.sol';
+import '@jbx-protocol-v2/contracts/interfaces/IJBToken.sol';
+import '@jbx-protocol-v2/contracts/interfaces/IJBTokenStore.sol';
+import '@jbx-protocol-v1/contracts/interfaces/ITicketBooth.sol';
+import '@openzeppelin/contracts/access/Ownable.sol';
+
+/** 
+  @notice
+  An ERC-20 token that can be used by a project in the `JBTokenStore` & also this takes care of the migration of the v1 & v2 project tokens for v3.
+
+  @dev
+  Adheres to -
+  IJBToken: Allows this contract to be used by projects in the JBTokenStore.
+
+  @dev
+  Inherits from -
+  ERC20Votes: General token standard for fungible membership with snapshot capabilities sufficient to interact with standard governance contracts. 
+  Ownable: Includes convenience functionality for checking a message sender's permissions before executing certain transactions.
+*/
+contract JBV3Token is ERC20Permit, Ownable, IJBToken {
+  //*********************************************************************//
+  // --------------------------- custom errors ------------------------- //
+  //*********************************************************************//
+  error BAD_PROJECT();
+
+
+  //*********************************************************************//
+  // --------------------- public stored properties -------------------- //
+  //*********************************************************************//
+
+  /** 
+    @notice
+    The ID of the project that this token should be exclusively used for. Send 0 to support any project. 
+  */
+  uint256 public immutable projectId;
+
+  /** 
+    @notice
+    The V1 Token Booth Instance. 
+  */
+  ITicketBooth public immutable v1TicketBooth;
+
+  /** 
+    @notice
+    The V2 Token Store Instance. 
+  */
+  IJBTokenStore public immutable v2TokenStore;
+
+  /** 
+    @notice
+    Storing the id's to migrate from for the v3 project ID. 
+  */
+  mapping(uint256 => MigrationInfo) public migrationOf;
+
+  /** 
+    @notice
+    v2ProjectId to migrate from. 
+    v1ProjectId to migrate from. 
+  */
+  struct MigrationInfo {
+    uint128 v2ProjectId;
+    uint128 v1ProjectId;
+  }
+
+  //*********************************************************************//
+  // ------------------------- external views -------------------------- //
+  //*********************************************************************//
+
+  /** 
+    @notice
+    The total supply of this ERC20.
+
+    @param _projectId the ID of the project to which the token belongs. This is ignored.
+
+    @return The total supply of this ERC20, as a fixed point number.
+  */
+  function totalSupply(uint256 _projectId) external view override returns (uint256) {
+    _projectId; // Prevents unused var compiler and natspec complaints.
+
+    (uint128 v2ProjectId, uint128 v1ProjectId) = getMigrationInfo(projectId);
+
+    return super.totalSupply() + 
```

Comment:
this might be unchecked from an arithmetic pov (ie will never be <0 as the invariant balanceOf(X) <= totalSupply should/might/need to hold) but concerned by the overflow for really large supply... wdyt?

Reply:
I think for safety side let's keep it as it is just in case for projects with very high supply